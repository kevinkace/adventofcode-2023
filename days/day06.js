export function parse(str) {
    return str.split(/\r?\n/);
}

export class Games {
    constructor(str) {
        this.races = [];
        this.parseRaces(str);
    }

    parseRaces(str) {
        const [times, distances] = str.split(/\r?\n/).map(line => line.split(/\s+/).filter((_, i) => i));

        for (let raceIdx = 0; raceIdx < times.length; raceIdx++) {
            const time = times[raceIdx];
            const distance = distances[raceIdx];

            this.races.push({ time, distance });
        }
    }

    getNumberWinning() {
        const numberWinningHoldTimes = [];

        for (const race of this.races) {
            let winningHoldTimes = 0;

            for (let holdTime = 1; holdTime < race.time; holdTime++) {
                const speed = holdTime;
                const dist = speed * (race.time - holdTime);

                if (dist > race.distance) {
                    winningHoldTimes++;
                }
            }

            numberWinningHoldTimes.push(winningHoldTimes);
        }

        // console.log(numberWinningHoldTimes);

        return numberWinningHoldTimes;
    }

    getProductOfWinning() {

        return this.getNumberWinning().reduce((acc, cur) => acc * cur, 1);
    }

    toString() {
        return this.races.map((race, idx) => {
            return [ idx, race.time, race.distance  ].join(" ");
        }).join("\n");
    }
}

export class Games2 extends Games {
    parseRaces(str) {
        const [time, distance] = str
            .split(/\r?\n/)
            .map(line => {
                return  line.split(":")[1].replaceAll(" ", "");
            });


        this.races.push({ time, distance });
    }
}
