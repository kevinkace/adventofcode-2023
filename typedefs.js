/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {Round[]} rounds
 */

/**
 * @typedef {Object} Round
 * @property {number} round
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

export default {};
