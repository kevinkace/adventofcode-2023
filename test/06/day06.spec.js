import { readFile }      from "node:fs/promises";
import { fileURLToPath } from "node:url";
import path              from "node:path";

import { expect, test, describe } from "vitest";

import { Games, Games2 } from "../../days/day06.js";

import typedefs from "../../typedefs.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const sample = await readFile(path.join(__dirname, "sample.txt"), "utf8");
const input  = await readFile(path.join(__dirname, "input.txt"), "utf8");


describe("day 06", () => {
    test("can parse input", () => {
        const sampleRaces = new Games(sample);
        const races       = new Games(input);

        const sampleString = sampleRaces.toString();

        expect(sampleString).toMatchSnapshot();
        expect(races.toString()).toMatchSnapshot();
    });

    test("can get winning", () => {
        const sampleRaces = new Games(sample);

        expect(sampleRaces.getNumberWinning()).toMatchSnapshot();
    });

    test("product of number winning", () => {
        const sampleRaces = new Games(sample);

        const productOfWinning = sampleRaces.getProductOfWinning();

        expect(productOfWinning).toBe(288);
    });

    test("answer p1", () => {
        const races = new Games(input);

        const productOfWinning = races.getProductOfWinning();

        expect(productOfWinning).toBe(1195150);
    });

    test("can parse v2", () => {
        const sampleRaces = new Games2(sample);

        expect(sampleRaces.getProductOfWinning()).toBe(71503);
    });

    // this takes a long time!!!
    test("answer p2", () => {
        const races = new Games2(input);

        const productOfWinning = races.getProductOfWinning();

        expect(races.races).toMatchSnapshot();

        // console.log({productOfWinning})

        expect(productOfWinning).toBe(42550411);
    });
});
