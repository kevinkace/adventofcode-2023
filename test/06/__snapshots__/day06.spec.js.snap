// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 06 > answer p2 1`] = `
[
  {
    "distance": "302147610291404",
    "time": "54946592",
  },
]
`;

exports[`day 06 > can get winning 1`] = `
[
  4,
  8,
  9,
]
`;

exports[`day 06 > can parse input 1`] = `
"0 7 9
1 15 40
2 30 200"
`;

exports[`day 06 > can parse input 2`] = `
"0 54 302
1 94 1476
2 65 1029
3 92 1404"
`;
